### DISCUSSION TITLE
Calling custom function externally

### DISCUSSION THREAD

**Discussion by haithium** [qtile/qtile#5028] [Q&A] [answered]
Hi, I'm trying to create a dynamic TextBox widget and manipulate it using the Commands API. The problem I'm facing is that I want to pass the string from the `text` parameter in the TextBox as an argument to a function that returns a different string value. However, The behavior is inconsistent - the function always returns false, yet it gives the same value that should be true.

```py
def evaluate_text(text):
    return "good" if text == "abc" else text
  
widget.TextBox(
      text="abc",
      fmt=evaluate_text("{}") # returns the same text value (abc)
  ),
```

Is there a way to achieve this or call a custom function within the Qtile configuration using the Commands API?

**Comment by elParaguayo** [comment]
I'll be honest. I've got no idea what your code is trying to achieve but I think there are a few misunderstandings in your code which may be part of your issue:
1) That `evaluate_text` function is only called once when the widget is first initialised.
2) It's being passed with a value of `"{}"` (which doesn't equal `"abc"`)
3) The `fmt` argument of the widget is a way to format text _after_ the text has been set.
4) `fmt` takes  a format string where `{}` is replaced by the text of the widget.

If you want to update a widget via the commands API, there are a few ways to do it:
* First, give the widget a `name` parameter when creating it as this allows you to identify it easily via the API e.g. `widget.TextBox(name="mytextbox", ...)`
* Updating it can be done in a number of ways:
  * Key binding: `lazy.widget["mytextbox"].update("New text")`
  * In a function in your config: `qtile.widgets_map["mytextbox"].update("New text")` (remember to do `from libqtile import qtile` in your config)
  * In an external command: `qtile cmd-obj -o widget mytext box -f update -a "New text"`

Hope that helps.

**Reply by haithium** [reply]
Awesome! However, Is there a way to control it from the outside. Let's say I have a function that handles some actions based on an argument in my config, but the problem is that I can't call it from outside.
And it can't be assigned a keybinding because it can take different arguments, making it difficult to assign a keybinding for each one

**Reply by elParaguayo** [reply]
Can you explain more? I don't really understand how you want it to work. I.e. what's the event that triggers the need to change the text? Where does the new text value come from?

**Reply by haithium** [reply]
Sure, I have a `layer_layout` function that handles content inside a TextBox. Based on an argument, it renders the layout like the image below. I want to call it from outside the config because I want to integrate it with my keyboard config and other tools, but I don't know how to call it externally.
![image](https://github.com/user-attachments/assets/732d3e20-bf89-40bd-bec8-bab750538146)

```py
layers = {
    "layer_name1": "<span fgcolor='#45475a'>q</span>...",
    "layer_name2": "<span fgcolor='#89b4fa'>1</span>...",
    # ...
}


def layer_layout(layer_name):
    if layer_name in layers:
        layout = layers[layer_name]
        return qtile.widgets_map["which-layer"].update(layout)

    return qtile.widgets_map["which-layer"].update(f"{layer_name} not found.")
```

```py
 widget.TextBox(text=" ", name="which-layer")
```

**Reply by elParaguayo** [reply, marked as answer]
You probably want to use the `user` hook. See https://docs.qtile.org/en/stable/manual/ref/hooks.html#libqtile.hook.subscribe.user

**Reply by haithium** [reply]
@elParaguayo Awesome! Thanks for your help.
